pkg/server/event_handler.go

```diff
@@ -234,33 +240,52 @@ func (h *EventHandler[T, P]) HTTPHandler() http.Handler {
 // [GCS notification]: https://cloud.google.com/storage/docs/pubsub-notifications#format
 func (h *EventHandler[T, P]) Handle(ctx context.Context, m pubsub.Message) error {
 	logger := logging.FromContext(ctx)
+
+	eventBytes, err := h.generatePmapEventBytes(ctx, m)
+
+	attr := map[string]string{}
+
+	if err != nil {
+		if !pmaperrors.Is(err) {
+			return err
+		}
+		attr[AttrKeyNonRetryableErr] = err.Error()
+		logger.Errorw(err.Error(), "bucketId", m.Attributes["bucketId"], "objectId", m.Attributes["objectId"])
+		if err := h.failureMessenger.Send(ctx, eventBytes, attr); err != nil {
+			return fmt.Errorf("failed to send failure event downstream: %w", err)
+		}
+		return nil
+	}
+	if err := h.successMessenger.Send(ctx, eventBytes, attr); err != nil {
+		return fmt.Errorf("failed to send succuss event downstream: %w", err)
+	}
+	return nil
+}
+
+func (h *EventHandler[T, P]) generatePmapEventBytes(ctx context.Context, m pubsub.Message) ([]byte, error) {
 	// Get the GCS object as a proto message given GCS notification information.
 	p, err := h.getGCSObjectProto(ctx, m.Attributes)
 	if err != nil {
-		return fmt.Errorf("failed to get GCS object: %w", err)
+		return nil, fmt.Errorf("failed to get GCS object: %w", err)
 	}
-
-	// TODO(#20): we need to have a way to differentiate retryable err vs. not.
-	// For non-retryable error, we need to have them enter a different BQ table per design.
-	// Currently all error events are sent to downstream if failureMessenger is provided
-	// including those retried events.
 	var processErr error
 	for _, processor := range h.processors {
 		if err := processor.Process(ctx, p); err != nil {
 			processErr = fmt.Errorf("failed to process object: %w", err)
 			break
 		}
 	}
+
 	payload, err := anypb.New(p)
 	if err != nil {
-		return fmt.Errorf("failed to convert object to pmap event payload: %w", err)
+		return nil, fmt.Errorf("failed to convert object to pmap event payload: %w", err)
 	}
 
 	var gr *v1alpha1.GitHubSource
 	if m.Attributes["payloadFormat"] == "JSON_API_V1" {
 		gr, err = parseGitHubSource(ctx, m.Data, m.Attributes)
 		if err != nil {
-			return fmt.Errorf("failed to parse metadata: %w", err)
+			return nil, fmt.Errorf("failed to parse metadata: %w", err)
```

Comment:
Should also join the `processErr` here. And add a comment (see the one below for an example)